src/scheduler/scheduler_test.cpp

```diff
@@ -0,0 +1,461 @@
+  }
+
+ private:
+  SamplingParameter sampling_param_;
+  StoppingCriteria stopping_criteria_;
+  Tokenizer* tokenizer_;
+
+  std::map<std::string, SequenceResult> outputs_;
+  bool is_finished = false;
+};
+
+TEST(SpeculativeSchedulerTest, Speculative4StepsPartiallyMatchTest) {
+  // who[1058] is[338] messi[4473] ?[29973]
+  const std::vector<int64_t> spec_token_ids = {1058, 338, 4473, 29973};
+  const std::vector<int64_t> valid_token_ids = {1058, 338, 4473, 1058, 338, 4473};
+  const uint64_t spec_steps = 4;
+
+  TestableSpeculativeScheduler scheduler(spec_token_ids, valid_token_ids,
+      spec_steps);
+  RequestWrapper request_wrapper(scheduler.tokenizer());
+
+  std::vector<int> input_tokens = {1058};
+  // expect output tokens: [1058] + [1058, 338, 4473, 1058] 
+  // expect output string: who who is messi who
+  const std::string expect_output("who who is messi who");
+  auto request = request_wrapper.create_request(input_tokens);
+  scheduler.schedule(request);
+  scheduler.step(absl::Seconds(1));
+  absl::SleepFor(absl::Seconds(2));
+  auto sequence_result = request_wrapper.sequence_output();
+
+  EXPECT_EQ(spec_steps, scheduler.get_ssm_execute_model_calls());
+  EXPECT_EQ(0, scheduler.get_llm_execute_model_calls());
+  EXPECT_EQ(0, scheduler.get_ssm_validate_calls());
+  EXPECT_EQ(1, scheduler.get_llm_validate_calls());
+
+  EXPECT_EQ(sequence_result.finish_reason, FinishReason::NONE);
+  EXPECT_EQ(sequence_result.output_text, expect_output);
+}
+
+TEST(SpeculativeSchedulerTest, Speculative3StepsFullyMatchTest) {
+  // who[1058] is[338] messi[4473] ?[29973]
+  const std::vector<int64_t> spec_token_ids = {1058, 338, 4473};
+  const std::vector<int64_t> valid_token_ids = {1058, 338, 4473, 1058};
+  const uint64_t spec_steps = 3;
+
+  TestableSpeculativeScheduler scheduler(spec_token_ids, valid_token_ids,
+      spec_steps);
+  RequestWrapper request_wrapper(scheduler.tokenizer());
+
+  std::vector<int> input_tokens = {1058};
+  // expect output tokens: [1058] + [1058, 338, 4473, 1058] 
+  // expect output string: who who is messi who
+  const std::string expect_output("who who is messi who");
+  auto request = request_wrapper.create_request(input_tokens);
+  scheduler.schedule(request);
+  scheduler.step(absl::Seconds(1));
+  absl::SleepFor(absl::Seconds(2));
+  auto sequence_result = request_wrapper.sequence_output();
+
+  EXPECT_EQ(spec_steps, scheduler.get_ssm_execute_model_calls());
+  EXPECT_EQ(0, scheduler.get_llm_execute_model_calls());
+  EXPECT_EQ(0, scheduler.get_ssm_validate_calls());
+  EXPECT_EQ(1, scheduler.get_llm_validate_calls());
+
+  EXPECT_EQ(sequence_result.finish_reason, FinishReason::NONE);
+  EXPECT_EQ(sequence_result.output_text, expect_output);
+}
+
+TEST(SpeculativeSchedulerTest, Speculative4StepsNoMatchTest) {
+  // who[1058] is[338] messi[4473] ?[29973]
+  const std::vector<int64_t> spec_token_ids = {1058, 338, 4473, 29973};
+  const std::vector<int64_t> valid_token_ids = {338, 4473, 29973, 1058, 338};
```

Comment:
is it a valid scenario that the valid_token_ids is longer than spec_token_ids in one step?